### PR TITLE
Minor changes in Univariate polynomials

### PR DESCRIPTION
**Pull request by srajangarg**
- removed the unecessary `dict_add_term` function
- Also removed `SYMENGINE_ASSERT` from vector constructor, because the polynomial formed is guaranteed to be canonical
- UnivariateInt had an incorrect `is_canonical`, for the 0 polynomial
- minor formatting/style changes in `from_dict`
- `mul_poly` was initially constructing a redundant vector (which could be too inefficient for sparse polys) Now uses `from_dict`. Constructs the dictionary on the go

More minor changes to follow.
@isuruf @Sumith1896

### CODE REVIEW

**Comment by srajangarg** [comment]
CodeCov failure can be ignored.

File: symengine/polynomial.cpp [hunk 1]
```diff
@@ -20,11 +20,15 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(
     const RCP<const Symbol> &var, const std::vector<integer_class> &v)
     : var_{var}
 {
-    for (unsigned int i = 0; i < v.size(); i++)
-        if (v[i] != 0)
-            dict_add_term(dict_, v[i], i);
-    degree_ = v.size() - 1;
-    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Why is this change necessary?

File: symengine/polynomial.cpp [hunk 2]
```diff
@@ -90,15 +94,14 @@ RCP<const UnivariateIntPolynomial>
...
+    auto iter = d.begin();
+    while (iter != d.end()) {
+        if (iter->second == 0) {
+            auto toErase = iter;
+            iter++;
             d.erase(toErase);
-        } else {
-            itter++;
-        }
+        } else
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Keep the `{`

File: symengine/polynomial.cpp (hunk 1 above)
Comment (srajangarg) [inline symengine/polynomial.cpp]: It's not necessary, but having a function just for adding terms to the dictionary did not make sense. Also, checking if it was canonical was not needed again, the function ensures that the constructed polynomial is canonical.

File: symengine/polynomial.cpp (hunk 2 above)
Comment (srajangarg) [inline symengine/polynomial.cpp]: Will do. Anything else?

**Comment by srajangarg** [comment]
What's wrong with the CodeCov CI?

**Comment by isuruf** [comment]
Codecov is rolling out major changes these days and that's why we are seeing some bugs from their side. Ignore the codecov status for now

**Comment by srajangarg** [comment]
@isuruf can this be merged in? I'm working on variations of `mul_poly`, and will get merge conflicts later on.

File: symengine/polynomial.cpp [hunk 3]
```diff
@@ -20,11 +20,15 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(
...
-    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
+    dict_ = {};
+    unsigned int deg = 0;
+    for (unsigned int i = 0; i < v.size(); i++) {
+        if (v[i] != 0) {
+            dict_[i] = v[i];
+            deg = i;
+        }
+    }
+    degree_ = deg;
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Use `(--dict.end())->first`

**Comment by isuruf** [comment]
Left one comment. Otherwise looks good

File: symengine/polynomial.cpp (hunk 3 above)
Comment (srajangarg) [inline symengine/polynomial.cpp]: Any particular reason? I'll also have to add a check if the dictionary isn't empty too.

Comment (isuruf) [inline symengine/polynomial.cpp]: Right, then it's okay

**Comment by isuruf** [comment]
+1. Let's wait for tests to pass
